example_command.go

```diff
@@ -0,0 +1,60 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/urfave/cli/v2"
+)
+
+type exampleCommand struct {
+	ExampleFlag string
+}
+
+var exampleCommandName = "example"
+
+func newExampleCommand() *cli.Command {
+	command := &exampleCommand{}
+	// TODO: Start hacking here
+	return &cli.Command{
+		Name:   exampleCommandName,
+		Usage:  "Start example command",
+		Before: command.validate,
+		Action: command.execute,
+		After:  command.shutdown,
+		Flags: []cli.Flag{
+			&cli.StringFlag{Destination: &command.ExampleFlag, Name: "flag", EnvVars: envVars("EXAMPLE_FLAG"), Value: "foo", Usage: "a demonstration how to configure the command"},
+		},
+	}
+}
+
+func (c *exampleCommand) validate(context *cli.Context) error {
+	log := AppLogger(context).WithName(exampleCommandName)
+	log.V(1).Info("validating config")
+	if c.ExampleFlag == "" {
+		return fmt.Errorf("option cannot be empty: %s", "flag")
+	}
+	return nil
+}
+
+func (c *exampleCommand) execute(context *cli.Context) error {
+	log := AppLogger(context).WithName(exampleCommandName)
+	go func() {
+		// This part enables graceful shutdowns. Can be removed if not needed.
```

Comment:
Maybe the sample should include using a `WaitGroup` as this is usually the way to go for waiting for a shutdown?

Reply:
I've added a WaitGroup example. However, I'm not sure how os.Exit comes to play now. os.Exit causes immediate termination, deferred functions will not run anymore, so wg.Done() wouldn't help at all. Maybe just remove the os.Exit statements?

Reply:
The way to go for shutting down a Go app usually works like this:

```go
func (c *exampleCommand) execute(ctx *cli.Context) error {
	cancelCtx, cancel := context.WithCancel(ctx.Context)
	var done sync.WaitGroup
	done.Add(1)
	log := AppLogger(ctx).WithName(exampleCommandName)
	go func() {
		// Shutdown hook. Can be used to gracefully shutdown listeners or pre-shutdown cleanup.
		defer done.Done()
		<-cancelCtx.Done()
		err := c.shutdown(ctx)
		if err != nil {
			log.Error(err, "cannot properly shut down")
			os.Exit(1)
		}
	}()
	go func() {
		defer cancel()
		time.Sleep(30 * time.Second)
	}()
	log.Info("Hello from example command!", "config", c)
	done.Wait()
	return nil
}
```

You don't need a `os.Exit(0)` a GO program exits as soon as the main Goroutine is done.

Reply:
Note this sample is not always true either. It depends it the shutdown block until done or the "server". The `defer done.Done()` has to be adjusted accordingly.

Reply:
More realistic:

```go
func (c *exampleCommand) execute(ctx *cli.Context) error {
	cancelCtx, cancel := context.WithCancel(ctx.Context)

	s := fakeServer{}

	var done sync.WaitGroup
	done.Add(2)
	log := AppLogger(ctx).WithName(exampleCommandName)
	go func() {
		// Shutdown hook. Can be used to gracefully shutdown listeners or pre-shutdown cleanup.
		defer done.Done()

		<-cancelCtx.Done()
		s.Shutdown()
	}()

	go func() {
		defer done.Done()
		defer cancel()
		s.Listen()
	}()
	log.Info("Hello from example command!", "config", c)
	done.Wait()
	return nil
}

type fakeServer struct {
	context.Context
	cancel context.CancelFunc
}

// Listen blocks for 30 seconds or until shut down
func (s *fakeServer) Listen() {
	s.Context, s.cancel = context.WithCancel(context.TODO())
	go func() {
		defer s.cancel()
		time.Sleep(10 * time.Second)
	}()
	<-s.Context.Done()
}

// Shutdown shuts down the server
func (s *fakeServer) Shutdown() {
	s.cancel()
}

```

Reply:
Thanks for the sample. However, I'm not sure I want to go ahead and assume that everything's somehow a server. I wanted to keep the example small enough so it can also run as a longer self-terminating job.

would you be fine with adding a comment that the sample is incomplete in its shutdown, but should demonstrate the possibilities?

Reply:
> would you be fine with adding a comment that the sample is incomplete in its shutdown, but should demonstrate the possibilities?

Yes that would be enough!